rust/datafusion/src/physical_plan/type_coercion.rs

```diff
@@ -69,13 +69,42 @@ pub fn data_types(
     signature: &Signature,
 ) -> Result<Vec<DataType>> {
     let valid_types = match signature {
-        Signature::Variadic(valid_types) => valid_types
+        Signature::Any(number) => {
+            if current_types.len() != *number {
+                return Err(DataFusionError::Plan(format!(
+                    "The function expected {} arguments but received {}",
+                    number,
+                    current_types.len()
+                )));
+            }
+            vec![(0..*number).map(|i| current_types[i].clone()).collect()]
+        }
+        Signature::Exact(valid_types) => vec![valid_types.clone()],
+        Signature::Uniform(valid_types) => {
+            let valid_signature = valid_types
+                .iter()
+                .filter(|x| x.len() == current_types.len())
+                .collect::<Vec<_>>();
+            if valid_signature.len() != 1 {
+                return Err(DataFusionError::Plan(format!(
+                    "The function expected {} arguments but received {}",
+                    valid_types
+                        .iter()
+                        .map(|x| x.len().to_string())
+                        .collect::<Vec<_>>()
+                        .join(" or "),
+                    current_types.len()
+                )));
+            }
+            cartesian_product(valid_signature.first().unwrap())
```

Comment:
Wont this coerce any type to the first variant, even if the latter variant is accepted?

I.e. if we use

```
Uniform(vec![
    vec![vec![A]],
    vec![vec![B]],
])
```

and pass arg types `vec![B]`, I would expect that no coercion would happen, but I suspect that this will coerce `B` to `A`, because the first entry with the same number of arguments is `vec![vec![A]]`.

Reply:
I suggest that we PR this separately with a single function that requires this type of signature, as we need to get this requires much more care than the other parts of this PR as it affects all future functions that use it.

Reply:
Thanks @jorgecarleitao . Yes I will split this out. 

A good example is lpad which is either:
[string, int] or [string, int, string]. I am away a couple of days but will split this out so we can work throught methodically.

Reply:
@jorgecarleitao as above: https://github.com/seddonm1/arrow/tree/oneof-function-signature